pkgs/build-support/replace-dependencies.nix

```diff
@@ -0,0 +1,140 @@
+{ lib, runCommandLocal, replaceDirectDependencies }:
+
+# Replace some dependencies in the requisites tree of drv, propagating the change all the way up the tree, even within other replacements, without a full rebuild.
+# This can be useful, for example, to patch a security hole in libc and still use your system safely without rebuilding the world.
+# This should be a short term solution, as soon as a rebuild can be done the properly rebuilt derivation should be used.
+# Each old dependency and the corresponding new dependency MUST have the same-length name, and ideally should have close-to-identical directory layout.
+#
+# Example: safeFirefox = replaceDependencies {
+#   drv = firefox;
+#   replacements = [
+#     {
+#       oldDependency = glibc;
+#       newDependency = glibc.overrideAttrs (oldAttrs: {
+#         patches = oldAttrs.patches ++ [ ./fix-glibc-hole.patch ];
+#       });
+#     }
+#     {
+#       oldDependency = libwebp;
+#       newDependency = libwebp.overrideAttrs (oldAttrs: {
+#         patches = oldAttrs.patches ++ [ ./fix-libwebp-hole.patch ];
+#       });
+#     }
+#   ];
+# };
+# This will first rebuild glibc and libwebp with your security patches.
+# Then it copies over firefox (and all of its dependencies) without rebuilding further.
+# In particular, the glibc dependency of libwebp will be replaced by the patched version as well.
+#
+# In rare cases, it is possible for the replacement process to cause breakage (for example due to checksum mismatch).
+# The cutoffPackages argument can be used to exempt the problematic packages from the replacement process.
+{ drv, replacements, cutoffPackages ? [ ], verbose ? true }:
+
+let
+  inherit (builtins) unsafeDiscardStringContext appendContext;
+  inherit (lib)
+    trace listToAttrs isStorePath readFile attrValues mapAttrs filter hasAttr
+    mapAttrsToList;
+  inherit (lib.attrsets) mergeAttrsList;
+
+  toContextlessString = x: unsafeDiscardStringContext (toString x);
+  warn = if verbose then trace else (x: y: y);
+
+  referencesOf = drv:
+    import (runCommandLocal "references.nix" {
+      exportReferencesGraph = [ "graph" drv ];
+    } ''
+      (echo {
+      while read path
+      do
+          echo "  \"$path\" = ["
+          read count
+          read count
+          while [ "0" != "$count" ]
+          do
+              read ref_path
+              if [ "$ref_path" != "$path" ]
+              then
+                  echo "    \"$ref_path\""
+              fi
+              count=$(($count - 1))
+          done
+          echo "  ];"
+      done < graph
+      echo }) > $out
+    '').outPath;
+
+  targetDerivations = [ drv ]
+    ++ map ({ newDependency, ... }: newDependency) replacements;
```

Comment:
Why do we need to the original `drv` here? It does not look like a "target" derivation, but rather like a "source" derivation...

Reply:
What I want here (and what the code correctly does) are the (root) derivations that will be needed for building the result, but before the rewrite. This variable was previously called `knownDerivations`, which I think was even worse. I'd be glad if you can come up with an actually good name, because I can't.